File: backend/apps/website_settings/schemas.py
from typing import Literal
from uuid import UUID

from ninja import Schema
from pydantic import Field


SettingValueTypeValue = Literal[
    "string",
    "text",
    "integer",
    "decimal",
    "boolean",
    "json",
    "email",
    "url",
    "color",
    "media",
]

SettingEnvironmentValue = Literal[
    "global",
    "development",
    "staging",
    "production",
]


class WebsiteSettingGroupCreateSchema(Schema):
    name: str
    slug: str
    description: str = ""
    icon: str = ""
    is_active: bool = True
    sort_order: int = 0


class WebsiteSettingGroupUpdateSchema(
    WebsiteSettingGroupCreateSchema
):
    pass


class WebsiteSettingCreateSchema(Schema):
    group_id: UUID
    key: str
    label: str
    description: str = ""
    value_type: SettingValueTypeValue = "string"
    environment: SettingEnvironmentValue = "global"
    value: str = ""
    json_value: object = Field(default_factory=dict)
    media_asset_id: UUID | None = None
    default_value: str = ""
    validation_rules: dict = Field(default_factory=dict)
    is_public: bool = False
    is_editable: bool = True
    is_required: bool = False
    is_active: bool = True
    sort_order: int = 0


class WebsiteSettingUpdateSchema(
    WebsiteSettingCreateSchema
):
    pass


class WebsiteSettingBulkItemSchema(Schema):
    setting_id: UUID
    value: str | None = None
    json_value: object | None = None
    media_asset_id: UUID | None = None


class WebsiteSettingBulkUpdateSchema(Schema):
    settings: list[
        WebsiteSettingBulkItemSchema
    ] = Field(default_factory=list)


class WebsiteSettingSchema(Schema):
    id: UUID
    group_id: UUID
    group_name: str
    key: str
    label: str
    description: str
    value_type: str
    environment: str
    value: str
    json_value: object
    media_asset_id: UUID | None
    default_value: str
    validation_rules: dict
    typed_value: object
    is_public: bool
    is_editable: bool
    is_required: bool
    is_active: bool
    sort_order: int


class WebsiteSettingGroupSchema(Schema):
    id: UUID
    name: str
    slug: str
    description: str
    icon: str
    is_active: bool
    sort_order: int
    setting_count: int
    settings: list[WebsiteSettingSchema]


class PublicWebsiteSettingsSchema(Schema):
    environment: str
    settings: dict

File: apps/dashboard/src/features/website-settings/schemas.ts
import {
  z,
} from "zod";

import {
  settingEnvironments,
  settingValueTypes,
} from "./types";

const jsonObjectSchema = z.record(
  z.string(),
  z.unknown(),
);

export const websiteSettingSchema =
  z.object({
    id: z.string().uuid(),
    group_id: z.string().uuid(),
    group_name: z.string(),
    key: z.string(),
    label: z.string(),
    description: z.string(),
    value_type: z.enum(
      settingValueTypes,
    ),
    environment: z.enum(
      settingEnvironments,
    ),
    value: z.string(),
    json_value: z.unknown(),
    media_asset_id:
      z.string().uuid().nullable(),
    default_value: z.string(),
    validation_rules:
      jsonObjectSchema,
    typed_value: z.unknown(),
    is_public: z.boolean(),
    is_editable: z.boolean(),
    is_required: z.boolean(),
    is_active: z.boolean(),
    sort_order: z.number().int(),
  });

export const websiteSettingGroupSchema:
z.ZodType<{
  id: string;
  name: string;
  slug: string;
  description: string;
  icon: string;
  is_active: boolean;
  sort_order: number;
  setting_count: number;
  settings: z.infer<
    typeof websiteSettingSchema
  >[];
}> = z.object({
  id: z.string().uuid(),
  name: z.string(),
  slug: z.string(),
  description: z.string(),
  icon: z.string(),
  is_active: z.boolean(),
  sort_order: z.number().int(),
  setting_count: z.number().int(),
  settings: z.array(
    websiteSettingSchema,
  ),
});

export const websiteSettingGroupPayloadSchema =
  z.object({
    name: z.string().trim().min(
      1,
      "Group name is required.",
    ),
    slug: z.string().trim().min(
      1,
      "Group slug is required.",
    ),
    description: z.string(),
    icon: z.string(),
    is_active: z.boolean(),
    sort_order: z.number().int(),
  });

export const websiteSettingPayloadSchema =
  z.object({
    group_id: z.string().uuid(),
    key: z.string().trim().min(
      1,
      "Setting key is required.",
    ),
    label: z.string().trim().min(
      1,
      "Setting label is required.",
    ),
    description: z.string(),
    value_type: z.enum(
      settingValueTypes,
    ),
    environment: z.enum(
      settingEnvironments,
    ),
    value: z.string(),
    json_value: z.unknown(),
    media_asset_id:
      z.string().uuid().nullable(),
    default_value: z.string(),
    validation_rules:
      jsonObjectSchema,
    is_public: z.boolean(),
    is_editable: z.boolean(),
    is_required: z.boolean(),
    is_active: z.boolean(),
    sort_order: z.number().int(),
  });

export const publicWebsiteSettingsSchema =
  z.object({
    environment: z.string(),
    settings: jsonObjectSchema,
  });
